src/levels.rs

```diff
@@ -1,5 +1,1053 @@
+                if !skip_key.is_empty() {
+                    if same_key(&iter_key, &skip_key) {
+                        num_skips += 1;
+                        // TODO: update stats of vlog
+
+                        iter.next();
+                        continue;
+                    } else {
+                        skip_key.clear();
+                    }
+                }
+
+                if !same_key(&iter_key, &last_key) {
+                    if let KeyRange::Range { left: _, right } = kr {
+                        if !right.is_empty()
+                            && COMPARATOR.compare_key(&iter_key, right) != std::cmp::Ordering::Less
+                        {
+                            break;
+                        }
+                    }
+
+                    // Ensure that all versions of the key are stored in the same
+                    // sstable, and not divided across multiple tables at the same
+                    // level.
+                    if builder.reach_capacity() {
+                        break;
+                    }
+
+                    last_key.clear();
+                    last_key.extend_from_slice(iter.key());
+                    num_versions = 0;
+
+                    match &mut table_kr {
+                        KeyRange::Range { left, right } => {
+                            if left.is_empty() {
+                                *left = Bytes::copy_from_slice(&iter_key);
+                            }
+                            *right = Bytes::copy_from_slice(&last_key);
+                        }
+                        KeyRange::Inf => todo!(),
+                        KeyRange::Empty => {
+                            table_kr = KeyRange::Range {
+                                left: Bytes::copy_from_slice(&iter_key),
+                                right: Bytes::copy_from_slice(&last_key),
+                            }
+                        }
+                    }
+
+                    // TODO: Do range check.
+                }
+
+                let vs = iter.value();
+                let version = get_ts(&iter_key);
+
+                if version <= discard_ts && vs.meta & crate::value::VALUE_MERGE_ENTRY == 0 {
+                    num_versions += 1;
+
+                    let last_valid_version = vs.meta & crate::value::VALUE_DISCARD_EARLIER_VERSIONS
+                        != 0
+                        || num_versions == self.opts.num_versions_to_keep;
+
+                    let is_expired = is_deleted_or_expired(vs.meta, vs.expires_at);
+
+                    if is_expired || last_valid_version {
+                        skip_key = BytesMut::from(&iter_key[..]);
+                        #[allow(clippy::if_same_then_else)]
+                        if !is_expired && last_valid_version {
+                            // do nothing
+                        } else if has_overlap {
+                            // do nothing
+                        } else {
```

Comment:
How about this key has overlap with levels below next level?

Reply:
> How about this key has overlap with levels below next level?

Sorry I misunderstood what you were asking. As for the current key, it will be added to the table, although it is deleted or expired.